src/matrices.jl

```diff
@@ -0,0 +1,66 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENCE in the project root.
+# ------------------------------------------------------------------
+
+"""
+    UnitMatrix{T}(d)
+    UnitColumn{T}(d)
+    F{T}(d)
+    G{T}(N)
+    H{T}(d)
+
+Elementary matrices as defined by Aitchison 1986.
+
+## Examples
+
+```julia
+julia> UnitMatrix(5)
+julia> UnitColumn{Int}(2)
+julia> G(4)
+julia> H{Float64}(2)
+```
+
+When the type of the elements isn't explicitly defined, `Float64` is used.
+"""
+
+struct UnitMatrix{T} <: AbstractArray{T, 2}
+    d::Int
+end
+UnitMatrix(d::Int) = UnitMatrix{Float64}(d)
+Base.size(J::UnitMatrix) = (J.d, J.d)
+Base.IndexStyle(::Type{<:UnitMatrix}) = IndexLinear()
+Base.getindex(J::UnitMatrix, i::Int, j::Int) = convert(eltype(J), 1)
+
+
+struct UnitColumn{T} <: AbstractArray{T, 1}
+    d::Int
+end
+UnitColumn(d::Int) = UnitColumn{Float64}(d)
+Base.size(j::UnitColumn) = (j.d,)
+Base.IndexStyle(::Type{UnitColumn}) = IndexLinear()
+Base.getindex(j::UnitColumn, i::Int) = convert(eltype(j), 1)
+
+
+struct F{T} <: AbstractArray{T, 2}
+    d::Int
+end
+F(d::Int) = F{Float64}(d)
+Base.size(A::F) = (A.d, A.d+1)
+Base.IndexStyle(::Type{F}) = IndexLinear()
+Base.getindex(A::F, i::Int, j::Int) = convert(eltype(A), ifelse(i==j, 1, 0) - (j==(A.d+1)))
```

Comment:
Instead of using `convert(eltype(A), x)` you can use `one(T)` and `zero(T)` to get the unit and zero elements for the type T. This type T can be specified in the argument as `A::F{T} where {T}`. See the other suggestion above for a concrete example.